[automation] ScriptFileWatcher breaks Jython reloading on change

I don't know enough to be able to say whether this belongs here or on openhba-addons. I'm putting it here for now because the timing of reports of problems almost exactly coincided with the merging of PR #2708. I'm seeing lots of reports from Jython users that changes made to their .py files are no longer being picked up. They have to restart openHAB for changes to be detected and loaded. 

In a typical Jython install using the Helper Libraries, a user will place their files into $OH_CONF/automation/jsr223/python/personal and library files would go in $OH_CONF/automation/lib/python/personal. The changes made in #2708 appears to stop watching those subfolders.

Current snapshot does not reload at all with a hard error: 
`ScriptEngineManagerImpl] - ScriptEngine for language 'py' could not be created for identifier: file:/etc/openhab/automation/jsr223/python/personal/my_file.py`

Interestingly, the scripts load once at startup, but not on change.
If you want me to install a special bundle, no problem, just let me know the steps.

```
2022-02-08 18:35:13.386 [TRACE] [org.openhab.core.service.WatchQueueReader         ] - Modification event for /etc/openhab/automation/jsr223/python/personal/1livingroom_lights.py 
2022-02-08 18:35:13.388 [TRACE] [org.openhab.core.service.WatchQueueReader         ] - Scheduled processing of /etc/openhab/automation/jsr223/python/personal/1livingroom_lights.py
2022-02-08 18:35:13.390 [TRACE] [org.openhab.core.service.WatchQueueReader         ] - Modification event for /etc/openhab/automation/jsr223/python/personal/1livingroom_lights.py 
2022-02-08 18:35:13.392 [TRACE] [org.openhab.core.service.WatchQueueReader         ] - Scheduled processing of /etc/openhab/automation/jsr223/python/personal/1livingroom_lights.py
2022-02-08 18:35:13.395 [TRACE] [org.openhab.core.service.WatchQueueReader         ] - Modification event for /etc/openhab/automation/jsr223/python/personal/1livingroom_lights.py 
2022-02-08 18:35:13.397 [TRACE] [org.openhab.core.service.WatchQueueReader         ] - Scheduled processing of /etc/openhab/automation/jsr223/python/personal/1livingroom_lights.py
2022-02-08 18:35:13.399 [TRACE] [org.openhab.core.service.WatchQueueReader         ] - Modification event for /etc/openhab/automation/jsr223/python/personal/1livingroom_lights.py 
2022-02-08 18:35:13.400 [TRACE] [org.openhab.core.service.WatchQueueReader         ] - Scheduled processing of /etc/openhab/automation/jsr223/python/personal/1livingroom_lights.py
2022-02-08 18:35:13.407 [TRACE] [org.openhab.core.service.WatchQueueReader         ] - Modification event for /etc/openhab/automation/jsr223/python/personal/1livingroom_lights.py 
2022-02-08 18:35:13.408 [TRACE] [org.openhab.core.service.WatchQueueReader         ] - Scheduled processing of /etc/openhab/automation/jsr223/python/personal/1livingroom_lights.py
2022-02-08 18:35:14.408 [TRACE] [org.openhab.core.service.WatchQueueReader         ] - Executing job for /etc/openhab/automation/jsr223/python/personal/1livingroom_lights.py
2022-02-08 18:35:14.410 [TRACE] [org.openhab.core.service.WatchQueueReader         ] - Job removed itself for /etc/openhab/automation/jsr223/python/personal/1livingroom_lights.py
2022-02-08 18:35:14.413 [DEBUG] [module.script.rulesupport.loader.ScriptFileWatcher] - Dequeued file:/etc/openhab/automation/jsr223/python/personal/1livingroom_lights.py
2022-02-08 18:35:14.415 [TRACE] [ion.module.script.internal.ScriptEngineManagerImpl] - scriptUnloaded() is not defined in the script
2022-02-08 18:35:14.461 [INFO ] [module.script.rulesupport.loader.ScriptFileWatcher] - Loading script '/etc/openhab/automation/jsr223/python/personal/1livingroom_lights.py'
2022-02-08 18:35:14.471 [ERROR] [ion.module.script.internal.ScriptEngineManagerImpl] - ScriptEngine for language 'py' could not be created for identifier: file:/etc/openhab/automation/jsr223/python/personal/1livingroom_lights.py
2022-02-08 18:35:14.473 [ERROR] [module.script.rulesupport.loader.ScriptFileWatcher] - Script loading error, ignoring file: /etc/openhab/automation/jsr223/python/personal/1livingroom_lights.py
```

One strange thing appeared in the log when I used the karaf update command. Here is the first snippet of the log, and an error that I've never seen before.

```
2022-02-14 21:27:01.736 [INFO ] [module.script.rulesupport.loader.ScriptFileWatcher] - Loading script '/etc/openhab/automation/jsr223/python/personal/1kitchen_lights.py'
2022-02-14 21:27:04.899 [ERROR] [ion.module.script.internal.ScriptEngineManagerImpl] - Error during evaluation of script 'file:/etc/openhab/automation/jsr223/python/personal/1kitchen_lights.py': TypeError: Error when calling the metaclass bases
    cannot create 'NoneType' instances in /etc/openhab/automation/jsr223/python/personal/1kitchen_lights.py at line number 3
2022-02-14 21:27:04.900 [DEBUG] [module.script.rulesupport.loader.ScriptFileWatcher] - Script loaded: /etc/openhab/automation/jsr223/python/personal/1kitchen_lights.py
2022-02-14 21:27:04.903 [DEBUG] [module.script.rulesupport.loader.ScriptFileWatcher] - Dequeued file:/etc/openhab/automation/jsr223/python/personal/1kitchen_lights.py
2022-02-14 21:27:04.903 [TRACE] [org.openhab.core.service.WatchQueueReader         ] - Executing job for /etc/openhab/automation/jsr223/python/personal/1studio_lights.py
2022-02-14 21:27:04.905 [TRACE] [org.openhab.core.service.WatchQueueReader         ] - Job removed itself for /etc/openhab/automation/jsr223/python/personal/1studio_lights.py
2022-02-14 21:27:04.908 [DEBUG] [module.script.rulesupport.loader.ScriptFileWatcher] - Enqueued file:/etc/openhab/automation/jsr223/python/personal/1kitchen_lights.py
2022-02-14 21:27:04.908 [DEBUG] [module.script.rulesupport.loader.ScriptFileWatcher] - Dequeued file:/etc/openhab/automation/jsr223/python/personal/7storage_lights.py
2022-02-14 21:27:04.912 [DEBUG] [module.script.rulesupport.loader.ScriptFileWatcher] - Enqueued file:/etc/openhab/automation/jsr223/python/personal/7storage_lights.py
2022-02-14 21:27:04.914 [DEBUG] [module.script.rulesupport.loader.ScriptFileWatcher] - Enqueued file:/etc/openhab/automation/jsr223/python/personal/7storage_lights.py
2022-02-14 21:27:04.916 [DEBUG] [module.script.rulesupport.loader.ScriptFileWatcher] - Dequeued file:/etc/openhab/automation/jsr223/python/personal/1presence.py
2022-02-14 21:27:04.919 [DEBUG] [module.script.rulesupport.loader.ScriptFileWatcher] - Enqueued file:/etc/openhab/automation/jsr223/python/personal/1presence.py
2022-02-14 21:27:04.919 [DEBUG] [module.script.rulesupport.loader.ScriptFileWatcher] - Dequeued file:/etc/openhab/automation/jsr223/python/personal/1studio_lights.py
2022-02-14 21:27:04.922 [DEBUG] [module.script.rulesupport.loader.ScriptFileWatcher] - Enqueued file:/etc/openhab/automation/jsr223/python/personal/1studio_lights.py
2022-02-14 21:27:05.159 [TRACE] [module.script.rulesupport.loader.ScriptFileWatcher] - Start level reached for file:/etc/openhab/automation/jsr223/python/personal/1bathroom.py
2022-02-14 21:27:05.161 [TRACE] [module.script.rulesupport.loader.ScriptFileWatcher] - ScriptEngineManager supports file:/etc/openhab/automation/jsr223/python/personal/1bathroom.py
2022-02-14 21:27:05.162 [INFO ] [module.script.rulesupport.loader.ScriptFileWatcher] - Loading script '/etc/openhab/automation/jsr223/python/personal/1bathroom.py'
2022-02-14 21:27:09.857 [DEBUG] [jsr223.jython.core.rules                          ] - Added rule 'BathroomLightFix'
...
```
After which it seems all files go on and load properly, including 1kitchen_lights.py.

The first 4 lines of 1kitchen_lights.py are these:
```
from core.actions import PersistenceExtensions
from core.actions import ScriptExecution
from core.rules import rule
from core.triggers import when
...
```

--- a/bundles/org.openhab.core.automation.module.script/src/main/java/org/openhab/core/automation/module/script/internal/ScriptEngineManagerImpl.java
+++ b/bundles/org.openhab.core.automation.module.script/src/main/java/org/openhab/core/automation/module/script/internal/ScriptEngineManagerImpl.java
@@ -54,9 +54,9 @@ import org.slf4j.LoggerFactory;
 @NonNullByDefault
 @Component(service = ScriptEngineManager.class)
 public class ScriptEngineManagerImpl implements ScriptEngineManager {
+    private static final String THREAD_POOL_NAME = "OH-scriptenginemanager";
 
-    private final ScheduledExecutorService scheduler = ThreadPoolManager
-            .getScheduledPool(ThreadPoolManager.THREAD_POOL_NAME_COMMON);
+    private final ScheduledExecutorService scheduler = ThreadPoolManager.getScheduledPool(THREAD_POOL_NAME);
 
     private final Logger logger = LoggerFactory.getLogger(ScriptEngineManagerImpl.class);
     private final Map<String, ScriptEngineContainer> loadedScriptEngineInstances = new HashMap<>();
@@ -222,6 +222,9 @@ public class ScriptEngineManagerImpl implements ScriptEngineManager {
             }
 
             if (scriptEngine instanceof AutoCloseable) {
+                // we cannot not use ScheduledExecutorService.execute here as it might execute the task in the calling
+                // thread (calling ScriptEngine.close in the same thread may result in a deadlock if the ScriptEngine
+                // tries to Thread.join)
                 scheduler.schedule(() -> {
                     AutoCloseable closeable = (AutoCloseable) scriptEngine;
                     try {
@@ -229,7 +232,7 @@ public class ScriptEngineManagerImpl implements ScriptEngineManager {
                     } catch (Exception e) {
                         logger.error("Error while closing script engine", e);
                     }
-                }, 1, TimeUnit.MILLISECONDS);
+                }, 0, TimeUnit.SECONDS);
             } else {
                 logger.trace("ScriptEngine does not support AutoCloseable interface");
             }

--- a/bundles/org.openhab.core.automation.module.script/src/main/java/org/openhab/core/automation/module/script/internal/ScriptEngineManagerImpl.java
+++ b/bundles/org.openhab.core.automation.module.script/src/main/java/org/openhab/core/automation/module/script/internal/ScriptEngineManagerImpl.java
@@ -20,6 +20,8 @@ import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
 
 import javax.script.Invocable;
 import javax.script.ScriptContext;
@@ -34,6 +36,7 @@ import org.openhab.core.automation.module.script.ScriptEngineContainer;
 import org.openhab.core.automation.module.script.ScriptEngineFactory;
 import org.openhab.core.automation.module.script.ScriptEngineManager;
 import org.openhab.core.automation.module.script.ScriptExtensionManagerWrapper;
+import org.openhab.core.common.ThreadPoolManager;
 import org.osgi.service.component.annotations.Activate;
 import org.osgi.service.component.annotations.Component;
 import org.osgi.service.component.annotations.Reference;
@@ -51,6 +54,9 @@ import org.slf4j.LoggerFactory;
 @NonNullByDefault
 @Component(service = ScriptEngineManager.class)
 public class ScriptEngineManagerImpl implements ScriptEngineManager {
+
+    private final ScheduledExecutorService scheduler = ThreadPoolManager
+            .getScheduledPool(ThreadPoolManager.THREAD_POOL_NAME_COMMON);
 
     private final Logger logger = LoggerFactory.getLogger(ScriptEngineManagerImpl.class);
     private final Map<String, ScriptEngineContainer> loadedScriptEngineInstances = new HashMap<>();
@@ -216,12 +222,14 @@ public class ScriptEngineManagerImpl implements ScriptEngineManager {
             }
 
             if (scriptEngine instanceof AutoCloseable) {
-                AutoCloseable closeable = (AutoCloseable) scriptEngine;
-                try {
-                    closeable.close();
-                } catch (Exception e) {
-                    logger.error("Error while closing script engine", e);
-                }
+                scheduler.schedule(() -> {
+                    AutoCloseable closeable = (AutoCloseable) scriptEngine;
+                    try {
+                        closeable.close();
+                    } catch (Exception e) {
+                        logger.error("Error while closing script engine", e);
+                    }
+                }, 1, TimeUnit.MILLISECONDS);
             } else {
                 logger.trace("ScriptEngine does not support AutoCloseable interface");
             }

--- a/bundles/org.openhab.core.automation.module.script/src/main/java/org/openhab/core/automation/module/script/internal/ScriptEngineManagerImpl.java
+++ b/bundles/org.openhab.core.automation.module.script/src/main/java/org/openhab/core/automation/module/script/internal/ScriptEngineManagerImpl.java
@@ -54,9 +54,9 @@ import org.slf4j.LoggerFactory;
 @NonNullByDefault
 @Component(service = ScriptEngineManager.class)
 public class ScriptEngineManagerImpl implements ScriptEngineManager {
-    private static final String THREAD_POOL_NAME = "OH-scriptenginemanager";
 
-    private final ScheduledExecutorService scheduler = ThreadPoolManager.getScheduledPool(THREAD_POOL_NAME);
+    private final ScheduledExecutorService scheduler = ThreadPoolManager
+            .getScheduledPool(ThreadPoolManager.THREAD_POOL_NAME_COMMON);
 
     private final Logger logger = LoggerFactory.getLogger(ScriptEngineManagerImpl.class);
     private final Map<String, ScriptEngineContainer> loadedScriptEngineInstances = new HashMap<>();

